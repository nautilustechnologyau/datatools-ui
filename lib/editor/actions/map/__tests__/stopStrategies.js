import clone from 'lodash/cloneDeep'

import { updateShapeDistTraveled } from '../stopStrategies'

const loopControlPoints = require('./fixtures/loop-ctrl-points.json')
const loopPatternSegments = require('./fixtures/loop-pattern-segments.json')
const loopPatternStops = require('./fixtures/loop-pattern-stops.json')

describe('editor > actions > stopStrategies', () => {
  describe('updateShapeDistTraveled', () => {
    it('should populate distance traveled in a loop pattern', () => {
      const clonedPatternStops = clone(loopPatternStops)
      updateShapeDistTraveled(loopControlPoints, loopPatternSegments, clonedPatternStops)

      expect(clonedPatternStops[0].shapeDistTraveled).toBe(0)

      // Traveled distance should be in strict ascending order.
      for (let i = 1; i < clonedPatternStops.length; i++) {
        expect(clonedPatternStops[i].shapeDistTraveled).toBeGreaterThan(clonedPatternStops[i - 1].shapeDistTraveled)
      }

      // The pattern contains a loop, so the total distance traveled from pattern start
      // should be populated as opposed to the distance when hitting a repeated stop the first time.
      expect(clonedPatternStops[23].stopId).toBe(clonedPatternStops[0].stopId)
      expect(clonedPatternStops[23].shapeDistTraveled).toBe(184921.39475283914)

      expect(clonedPatternStops[21].stopId).toBe(clonedPatternStops[2].stopId)
      expect(clonedPatternStops[2].shapeDistTraveled).toBe(1621.2549997010212)
      expect(clonedPatternStops[21].shapeDistTraveled).toBe(183802.09138428463)

      // See other individual computed traveled distances in snapshot.
      expect(clonedPatternStops).toMatchSnapshot()
    })
  })
})
