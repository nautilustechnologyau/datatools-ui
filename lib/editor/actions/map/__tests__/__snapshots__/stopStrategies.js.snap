// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`editor > actions > stopStrategies updateShapeDistTraveled should populate distance traveled in a loop pattern 1`] = `
Array [
  Object {
    "defaultTravelTime": 0,
    "id": 812,
    "shapeDistTraveled": 0,
    "stopId": "mni3",
    "stopSequence": 0,
  },
  Object {
    "defaultTravelTime": 720,
    "id": 813,
    "shapeDistTraveled": 1267.0907200881181,
    "stopId": "20",
    "stopSequence": 1,
  },
  Object {
    "defaultTravelTime": 180,
    "id": 814,
    "shapeDistTraveled": 1621.2549997010212,
    "stopId": "z745",
    "stopSequence": 2,
  },
  Object {
    "defaultTravelTime": 1800,
    "id": 815,
    "shapeDistTraveled": 27816.018654027193,
    "stopId": "56",
    "stopSequence": 3,
  },
  Object {
    "defaultTravelTime": 1440,
    "id": 816,
    "shapeDistTraveled": 54970.153868786,
    "stopId": "32",
    "stopSequence": 4,
  },
  Object {
    "defaultTravelTime": 180,
    "id": 817,
    "shapeDistTraveled": 57221.67317745581,
    "stopId": "mesm",
    "stopSequence": 5,
  },
  Object {
    "defaultTravelTime": 360,
    "id": 818,
    "shapeDistTraveled": 73839.4199301234,
    "stopId": "48",
    "stopSequence": 6,
  },
  Object {
    "defaultTravelTime": 780,
    "id": 819,
    "shapeDistTraveled": 90255.00980161014,
    "stopId": "41",
    "stopSequence": 7,
  },
  Object {
    "defaultTravelTime": 360,
    "id": 820,
    "shapeDistTraveled": 91772.97104090628,
    "stopId": "42",
    "stopSequence": 8,
  },
  Object {
    "defaultTravelTime": 300,
    "id": 821,
    "shapeDistTraveled": 93027.691348872,
    "stopId": "54",
    "stopSequence": 9,
  },
  Object {
    "defaultTravelTime": 780,
    "id": 822,
    "shapeDistTraveled": 108704.22121540575,
    "stopId": "106",
    "stopSequence": 10,
  },
  Object {
    "defaultTravelTime": 240,
    "id": 823,
    "shapeDistTraveled": 114069.642131669,
    "stopId": "2",
    "stopSequence": 11,
  },
  Object {
    "defaultTravelTime": 360,
    "id": 824,
    "shapeDistTraveled": 122591.76955559291,
    "stopId": "43",
    "stopSequence": 12,
  },
  Object {
    "defaultTravelTime": 360,
    "id": 825,
    "shapeDistTraveled": 123703.67475483804,
    "stopId": "104",
    "stopSequence": 13,
  },
  Object {
    "defaultTravelTime": 360,
    "id": 826,
    "shapeDistTraveled": 127934.07012854327,
    "stopId": "o6vb",
    "stopSequence": 14,
  },
  Object {
    "defaultTravelTime": 600,
    "id": 827,
    "shapeDistTraveled": 138359.3114221102,
    "stopId": "110",
    "stopSequence": 15,
  },
  Object {
    "defaultTravelTime": 600,
    "id": 829,
    "shapeDistTraveled": 148784.5527156771,
    "stopId": "95",
    "stopSequence": 17,
  },
  Object {
    "defaultTravelTime": 300,
    "id": 830,
    "shapeDistTraveled": 164871.9601365662,
    "stopId": "24",
    "stopSequence": 18,
  },
  Object {
    "defaultTravelTime": 240,
    "id": 831,
    "shapeDistTraveled": 181433.9825137495,
    "stopId": "31",
    "stopSequence": 19,
  },
  Object {
    "defaultTravelTime": 0,
    "id": 828,
    "shapeDistTraveled": 181975.28021083935,
    "stopId": "87",
    "stopSequence": 16,
  },
  Object {
    "defaultTravelTime": 180,
    "id": 832,
    "shapeDistTraveled": 183447.92710467172,
    "stopId": "20",
    "stopSequence": 20,
  },
  Object {
    "defaultTravelTime": 420,
    "id": 833,
    "shapeDistTraveled": 183802.09138428463,
    "stopId": "z745",
    "stopSequence": 21,
  },
  Object {
    "defaultTravelTime": 180,
    "id": 834,
    "shapeDistTraveled": 184825.47542408676,
    "stopId": "19",
    "stopSequence": 22,
  },
  Object {
    "defaultTravelTime": 60,
    "id": 835,
    "shapeDistTraveled": 184921.39475283914,
    "stopId": "mni3",
    "stopSequence": 23,
  },
]
`;
